File: test/analysis/project_index_test.rb
# frozen_string_literal: true

require_relative "../test_helper"
require_relative "../../lib/herb/analysis/project_index"

require "tmpdir"
require "fileutils"

class ProjectIndexTest < Minitest::Spec
  def setup
    @project_path = Dir.mktmpdir("herb_project_index_test")
    @view_root = File.join(@project_path, "app", "views", "posts")
    FileUtils.mkdir_p(@view_root)
  end

  def teardown
    FileUtils.rm_rf(@project_path)
  end

  def write(name, content)
    path = File.join(@view_root, name)
    FileUtils.mkdir_p(File.dirname(path))
    File.write(path, content)
    path
  end

  def indexed
    project = Herb::Analysis::ProjectIndex.new(@project_path)
    project.index_all
    project
  end

  def reindexed
    Herb::Analysis::ProjectIndex.new(@project_path).tap(&:index_all)
  end

  test "indexes partials and call sites" do
    write("index.html.erb", '<table><%= render "posts/row", post: post %></table>')
    row = write("_row.html.erb", "<tr></tr>")

    project = indexed

    assert_equal 1, project.partials.size
    assert_equal 1, project.graph.size
    assert_equal :always, project.graph.context_of(row).ancestor_verdict([], "table")
  end

  test "re-analyzes only the changed template" do
    entry = write("index.html.erb", '<html><body><table><%= render "posts/row", post: post %></table></body></html>')
    row = write("_row.html.erb", "<tr></tr>")

    project = indexed
    File.write(entry, '<html><body><form><%= render "posts/row", post: post %></form></body></html>')

    assert project.handle_change(entry)
    assert_equal :never, project.graph.context_of(row).ancestor_verdict([], "table")
    assert_equal :always, project.graph.context_of(row).ancestor_verdict([], "form")
  end

  test "lands in the same state as a full reindex after a change" do
    entry = write("index.html.erb", '<table><%= render "posts/row", post: post %></table>')
    write("_row.html.erb", "<tr></tr>")

    project = indexed
    File.write(entry, '<form><%= render "posts/row", post: post %></form>')
    project.handle_change(entry)

    assert_equal reindexed.graph.to_h, project.graph.to_h
  end

  test "accepts source for a template that has not been written to disk" do
    entry = write("index.html.erb", '<table><%= render "posts/row", post: post %></table>')
    row = write("_row.html.erb", "<tr></tr>")

    project = indexed
    project.handle_change(entry, '<form><%= render "posts/row", post: post %></form>')

    assert_equal :always, project.graph.context_of(row).ancestor_verdict([], "form")
  end

  test "picks up a partial added after the initial index" do
    entry = write("index.html.erb", '<table><%= render "posts/row", post: post %></table>')

    project = indexed

    assert_equal :unknown, project.graph.context_of(File.join(@view_root, "_row.html.erb")).ancestor_verdict([], "table")

    row = write("_row.html.erb", "<tr></tr>")
    project.handle_change(row)
    project.handle_change(entry)

    assert_equal :always, project.graph.context_of(row).ancestor_verdict([], "table")
    assert_equal reindexed.graph.to_h, project.graph.to_h
  end

  test "forgets a partial that was deleted" do
    write("index.html.erb", '<table><%= render "posts/row", post: post %></table>')
    row = write("_row.html.erb", "<tr></tr>")

    project = indexed

    assert project.remove(row)
    assert_equal 0, project.partials.size
  end

  test "forgets the call sites of a deleted template" do
    entry = write("index.html.erb", '<table><%= render "posts/row", post: post %></table>')
    row = write("_row.html.erb", "<tr></tr>")

    project = indexed

    assert project.remove(entry)
    assert_empty project.graph.callers_of(row)
  end

  test "reports no change for a file that is not a template" do
    write("index.html.erb", "<table></table>")

    project = indexed

    refute project.handle_change(File.join(@project_path, "README.md"))
  end

  test "tracks strict locals declared by a partial" do
    write("index.html.erb", '<%= render "posts/row", post: post %>')
    write("_row.html.erb", "<%# locals: (post:) %>\n<tr></tr>")

    declaration = indexed.partials.lookup("posts/row", nil)

    assert declaration.has_declaration
    assert_equal ["post"], declaration.required_locals
    refute declaration.accepts?("author")
  end

  test "exposes the view root it resolved" do
    write("index.html.erb", "<div></div>")

    assert_equal [File.join(@project_path, "app", "views")], indexed.view_roots.map(&:to_s)
  end
end

File: rust/herb-analysis/tests/project_index_test.rs
use std::fs;
use std::path::PathBuf;

use herb_analysis::project_index::ProjectIndex;
use herb_analysis::render_graph::Verdict;

struct Project {
  root: PathBuf,
}

impl Project {
  fn new(name: &str) -> Self {
    let root = std::env::temp_dir().join(format!("herb_project_index_rs_{}_{}", name, std::process::id()));

    let _ = fs::remove_dir_all(&root);
    fs::create_dir_all(root.join("app/views/posts")).expect("create project");

    Self { root }
  }

  fn write(&self, path: &str, source: &str) -> String {
    let full = self.root.join(path);

    fs::create_dir_all(full.parent().expect("parent")).expect("create dir");
    fs::write(&full, source).expect("write");

    full.to_str().expect("utf8").to_string()
  }

  fn indexed(&self) -> ProjectIndex {
    let mut project = ProjectIndex::new(&self.root);
    project.index_all();

    project
  }
}

impl Drop for Project {
  fn drop(&mut self) {
    let _ = fs::remove_dir_all(&self.root);
  }
}

#[test]
fn indexes_partials_and_call_sites() {
  let project = Project::new("index_all");
  project.write("app/views/posts/index.html.erb", r#"<table><%= render "posts/row" %></table>"#);
  let row = project.write("app/views/posts/_row.html.erb", "<tr></tr>");

  let indexed = project.indexed();

  assert_eq!(indexed.partials().expect("partials").size(), 1);
  assert_eq!(indexed.graph().expect("graph").size(), 1);
  assert_eq!(
    indexed.graph().expect("graph").context_of(&row).ancestor_verdict(&[], &["table"]),
    Verdict::Always
  );
}

#[test]
fn re_analyzes_only_the_changed_template() {
  let project = Project::new("change");
  let entry = project.write(
    "app/views/posts/index.html.erb",
    r#"<html><body><table><%= render "posts/row" %></table></body></html>"#,
  );
  let row = project.write("app/views/posts/_row.html.erb", "<tr></tr>");

  let mut indexed = project.indexed();

  fs::write(&entry, r#"<html><body><form><%= render "posts/row" %></form></body></html>"#).expect("rewrite");

  assert!(indexed.handle_change(&entry, None));

  let graph = indexed.graph().expect("graph");

  assert_eq!(graph.context_of(&row).ancestor_verdict(&[], &["table"]), Verdict::Never);
  assert_eq!(graph.context_of(&row).ancestor_verdict(&[], &["form"]), Verdict::Always);
}

#[test]
fn accepts_source_for_a_template_that_has_not_been_written_to_disk() {
  let project = Project::new("buffer");
  let entry = project.write("app/views/posts/index.html.erb", r#"<html><table><%= render "posts/row" %></table></html>"#);
  let row = project.write("app/views/posts/_row.html.erb", "<tr></tr>");

  let mut indexed = project.indexed();

  indexed.handle_change(&entry, Some(r#"<html><form><%= render "posts/row" %></form></html>"#));

  assert_eq!(
    indexed.graph().expect("graph").context_of(&row).ancestor_verdict(&[], &["form"]),
    Verdict::Always
  );
}

#[test]
fn picks_up_a_partial_added_after_the_initial_index() {
  let project = Project::new("added");
  let entry = project.write("app/views/posts/index.html.erb", r#"<html><table><%= render "posts/row" %></table></html>"#);

  let mut indexed = project.indexed();
  let row = project.write("app/views/posts/_row.html.erb", "<tr></tr>");

  indexed.handle_change(&row, None);
  indexed.handle_change(&entry, None);

  assert_eq!(
    indexed.graph().expect("graph").context_of(&row).ancestor_verdict(&[], &["table"]),
    Verdict::Always
  );
}

#[test]
fn forgets_a_partial_that_was_deleted() {
  let project = Project::new("removed");
  project.write("app/views/posts/index.html.erb", r#"<table><%= render "posts/row" %></table>"#);
  let row = project.write("app/views/posts/_row.html.erb", "<tr></tr>");

  let mut indexed = project.indexed();

  assert!(indexed.remove(&row));
  assert_eq!(indexed.partials().expect("partials").size(), 0);
}

#[test]
fn forgets_the_call_sites_of_a_deleted_template() {
  let project = Project::new("removed_caller");
  let entry = project.write("app/views/posts/index.html.erb", r#"<table><%= render "posts/row" %></table>"#);
  let row = project.write("app/views/posts/_row.html.erb", "<tr></tr>");

  let mut indexed = project.indexed();

  assert!(indexed.remove(&entry));
  assert!(indexed.graph().expect("graph").callers_of(&row).is_empty());
}

#[test]
fn reports_no_change_for_a_file_that_is_not_a_template() {
  let project = Project::new("not_template");
  project.write("app/views/posts/index.html.erb", "<table></table>");

  let mut indexed = project.indexed();
  let readme = project.root.join("README.md").to_str().expect("utf8").to_string();

  assert!(!indexed.handle_change(&readme, None));
}

#[test]
fn exposes_the_view_root_it_resolved() {
  let project = Project::new("view_root");
  project.write("app/views/posts/index.html.erb", "<div></div>");

  assert_eq!(project.indexed().view_roots().expect("view roots"), [project.root.join("app/views")]);
}
